lib/class-wp-theme-json-gutenberg.php

```diff
@@ -2859,19 +2865,17 @@ public function merge( $incoming ) {
 						}
 					}
 
-					if (
-						( 'theme' !== $origin ) ||
-						( 'theme' === $origin && $override_preset )
-					) {
-						_wp_array_set( $this->theme_json, $path, $content );
-					} else {
-						$slugs_node = static::get_default_slugs( $this->theme_json, $node['path'] );
-						$slugs      = array_merge_recursive( $slugs_global, $slugs_node );
+					// Filter out default slugs from theme presets when defaults should not be overridden.
+					if ( 'theme' === $origin && $prevent_override ) {
+						$slugs_node    = static::get_default_slugs( $this->theme_json, $node['path'] );
+						$preset_global = _wp_array_get( $slugs_global, $preset_metadata['path'], array() );
+						$preset_node   = _wp_array_get( $slugs_node, $preset_metadata['path'], array() );
+						$preset_slugs  = array_merge_recursive( $preset_global, $preset_node );
 
-						$slugs_for_preset = _wp_array_get( $slugs, $preset['path'], array() );
-						$content          = static::filter_slugs( $content, $slugs_for_preset );
-						_wp_array_set( $this->theme_json, $path, $content );
+						$content = static::filter_slugs( $content, $preset_slugs );
 					}
+
+					_wp_array_set( $this->theme_json, $path, $content );
```

Comment:
This is just a refactor so things look a little more clear. This is the process that I did to make sure the logic was exactly the same.

```php
// Initial NOT (!) because of else clause
// $override_preset === ! $prevent_override

! ( ( 'theme' !== $origin ) || ( 'theme' === $origin && ! $prevent_override ) )

! ( 'theme' !== $origin ) && ! ( 'theme' === $origin && ! $prevent_override )

'theme' === $origin && ! ( 'theme' === $origin && ! $prevent_override )

'theme' === $origin && ( 'theme' !== $origin || $prevent_override )

( 'theme' === $origin && 'theme' !== $origin ) || ( 'theme' === $origin && $prevent_override )

false || ( 'theme' === $origin && $prevent_override )

'theme' === $origin && $prevent_override
```

Both code paths ended in `_wp_array_set( $this->theme_json, $path, $content );` so that was moved out.